### PR TITLE
feat(DSG 2592, DSG 2603, DSG2631, DSG 2632, DSG 2608): Ability to view a conversation thread

### PR DESCRIPTION
**Pull request by thechechogomez**
# Basics

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Styling and layout changes
- [ ] Code style changes (formatting, local variables)
- [ ] Refactoring (no functional changes)
- [ ] Build or CI related changes
- [ ] Tests
- [ ] Other... Please describe:

## What changes does this PR introduce?

Ticket: https://dol-ewf.atlassian.net/browse/DSG-2592
Ticket: https://dol-ewf.atlassian.net/browse/DSG-2603
Ticket: https://dol-ewf.atlassian.net/browse/DSG-2631
Ticket: https://dol-ewf.atlassian.net/browse/DSG-2632
Ticket: https://dol-ewf.atlassian.net/browse/DSG-2608

It allows an agent to view and reply to a conversation. 

An agent can click on a conversation from the list of conversations and navigate to the specific conversation. On the conversation, a user can expand or collapse all messages, as well as individual messages. The agent can also reply to a conversation, and the message will be available to be viewed in the thread. 

## Screenshot of the affected changes

https://github.com/Nuvalence/dsgov-web/assets/60438413/630e164f-3e8f-45ea-88a1-44607b5b4810

https://github.com/Nuvalence/dsgov-web/assets/60438413/570b5ff5-5fcf-4405-a5b6-adfc2fada114

## Checklist

- [ ] I have updated the documentation accordingly
- [ ] I have confirmed all acceptance criteria defined by the story have been met
- [ ] I have confirmed that Sonar Qube findings for new code are reviewed and fixed, or will be whitelisted

### CODE REVIEW

File: libs/employer/feature/dashboard/src/lib/components/transaction-messages/transaction-messages.component.ts
Comment (davidyoung77) [inline libs/employer/feature/dashboard/src/lib/components/transaction-messages/transaction-messages.component.ts:1]: Should this live in the messaging library?

File: libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.ts [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+      uiLabel: 'Cancel',
+    },
+  ];
+
+  constructor(private readonly _route: ActivatedRoute, private readonly _loadingService: LoadingService, private readonly _messagingService: MessagingService) {
+    this.messages$ = this._fetchMessages
+      .asObservable()
+      .pipe(
+        delay(1), // This is needed to show the loading indicator
+        switchMap(() => this._route.paramMap),
```
Comment (davidyoung77) [inline libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.ts:105]: Maybe this should be switch map with loading

File: libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.ts [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        }),
+        tap(response => {
+          this.conversation = response.conversation;
+        }),
+      )
+      .pipe(
+        filter(response => !!response),
+        switchMap(response => combineLatest([this._replies, of(response)])),
+        map(([replies, response]) => [...replies, ...response.messages, response.conversation.originalMessage]),
+        tap((replies: ConversationMessageModel[]) => {
```
Comment (davidyoung77) [inline libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.ts:124]: The loading spinner won't wait for anything in the tap here just fyi, maybe this should happen in a switchMap?

File: libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.html [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      <h4 class="message__sender">{{message.senderDisplayName}}</h4>
+      <p class="message__timestamp">
+        {{message.timestamp | date: "MM/dd/yy, h:mm a"}}
+        <ng-container *ngIf="message.timestamp | nuverialTimeElapsed">({{message.timestamp | nuverialTimeElapsed}} ago)</ng-container>
+      </p>
+    </div>
+    <div class="message__collapsed" *ngIf="message.collapsed; else richText">
+      <p class="message__collapsed__text" #message_text>{{message.noFormatBody}}</p>
+      <!-- eslint-disable-next-line @angular-eslint/template/no-call-expression -->
+      <p class="message__collapsed__more" *ngIf="showMoreButton(message_text)" (click)="toggleCollapseMessage(message)" role="none">More</p>
```
Comment (davidyoung77) [inline libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.html:48]: Is this for lazy loading? Can we do this without disabling?

Comment (davidyoung77) [reply libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.html:48]: Does this only show up if there is more scroll beyond the element, I didn't notice this in the video

File: libs/shared/ui/nuverial/src/lib/pipes/time-elapsed/time-elapsed.pipe.ts
Comment (davidyoung77) [inline libs/shared/ui/nuverial/src/lib/pipes/time-elapsed/time-elapsed.pipe.ts:1]: Looks like this is almost a copy of our time remaining pipe. What would we need to do with the existing pipe to make it work for both scenarios?

**Review by davidyoung77 — APPROVED** [review]
Looks good, I have a few comments

File: libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.html (hunk 3 above)
Comment (thechechogomez) [reply libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.html:48]: Yeah. We need the disabling as we have to reference the element and call the function with it to know if the content could overflow the container and if it does it shows the `Show More` button. 

Screenshot 2024-02-08 at 11 03 14 AM

File: libs/employer/feature/dashboard/src/lib/components/transaction-messages/transaction-messages.component.ts
Comment (thechechogomez) [reply libs/employer/feature/dashboard/src/lib/components/transaction-messages/transaction-messages.component.ts:1]: I though about it, but much like the shell component, it is only a space for the router outlet to live. Not sure if the library should contain a component that only contains a breadcrumb and the router outlet.

File: libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.ts (hunk 1 above)
Comment (thechechogomez) [reply libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.ts:105]: The route params observable is almost instantaneous, don't really see the benefit of adding it early on and adding it to the current implementation breaks the loading spinner.

File: libs/shared/ui/nuverial/src/lib/pipes/time-elapsed/time-elapsed.pipe.ts
Comment (thechechogomez) [reply libs/shared/ui/nuverial/src/lib/pipes/time-elapsed/time-elapsed.pipe.ts:1]: We would need to specify if we want to reference agains "today" or against the date we pass to the pipe. Also we would either need to specify a format or hard code that if we reference agains the date we pass on we don't want the full message but only day or hours or minutes + seconds.

File: libs/employer/feature/dashboard/src/lib/components/transaction-messages/transaction-messages.component.ts
Comment (thechechogomez) [reply libs/employer/feature/dashboard/src/lib/components/transaction-messages/transaction-messages.component.ts:1]: If we were to move this to the conversation list, then if somewhere else we would need to display these messages without the breadcrumb we would have to refactor again.

**Review by davidyoung77 — APPROVED** [review]
Nice work

**Review by davidyoung77 — APPROVED** [review]

**Pull request merged**